include/jemalloc/internal/rtree_inlines.h

```diff
@@ -180,9 +188,9 @@ rtree_elm_acquire(tsdn_t *tsdn, rtree_t *rtree, rtree_ctx_t *rtree_ctx,
 JEMALLOC_INLINE extent_t *
 rtree_elm_read_acquired(tsdn_t *tsdn, const rtree_t *rtree, rtree_elm_t *elm) {
 	extent_t *extent;
-
-	assert(((uintptr_t)elm->pun & (uintptr_t)0x1) == (uintptr_t)0x1);
-	extent = (extent_t *)((uintptr_t)elm->pun & ~((uintptr_t)0x1));
+	void *ptr = atomic_load_p(&elm->child_or_extent, ATOMIC_RELAXED);
```

Comment:
1. They both compile down to a plain load instruction. Using an atomic does slightly restrict the optimizer, because it can't assume that redundant loads are guaranteed to return the same value any more. But in the cases where it's doing that, it would usually be introducing a bug:

```
void do_something(some_type *t) {
  void *alloc = NULL;
  bool b = t->should_alloc; /* Suppose this reads b = true */
  if (b) {
    alloc = do_alloc();
  }
  /*
   * Some code that doesn't change t->b. The optimizer decides not to put b in a register, but
   * does leave t in a register. To save itself from the useless stack-store, it doesn't materialize b
   * on the stack; it just reloads it from t.
   */
  /*
   * Suppose another thread sets b to false, and we pick up that modification. Because b never got
   * materialized on the stack, we'll see the updated value, and won't take the branch.
   */
  if (b) {
    /* This never runs; we leak. */
    do_free(alloc);
  }
}
```

2. Nope. The only thing it does is prevent that sort of compiler optimization. The only ordering guarantee we get is that there's a total ordering of the stores to a particular location, and our reads never go backwards in time in that ordering.

3. In this particular case, it's a type issue; atomic_u32_t and uint32_t are genuinely different types. In general it would be safe though; if the struct were:

```
struct elm_s {
  atomic_b_t lock;
  extent_t *value;
};
```

Then we can implement the lock just with elm->b, and do plain reads from elm->value safely with no races.